src/betterproto/plugin/models.py

```diff
@@ -44,81 +69,82 @@
+    FieldDescriptorProtoType.TYPE_FIXED32,  # 7
+    FieldDescriptorProtoType.TYPE_UINT32,  # 13
+    FieldDescriptorProtoType.TYPE_SFIXED32,  # 15
+    FieldDescriptorProtoType.TYPE_SFIXED64,  # 16
+    FieldDescriptorProtoType.TYPE_SINT32,  # 17
+    FieldDescriptorProtoType.TYPE_SINT64,  # 18
 )
-PROTO_BOOL_TYPES = (FieldDescriptorProto.TYPE_BOOL,)  # 8
-PROTO_STR_TYPES = (FieldDescriptorProto.TYPE_STRING,)  # 9
-PROTO_BYTES_TYPES = (FieldDescriptorProto.TYPE_BYTES,)  # 12
+PROTO_BOOL_TYPES = (FieldDescriptorProtoType.TYPE_BOOL,)  # 8
+PROTO_STR_TYPES = (FieldDescriptorProtoType.TYPE_STRING,)  # 9
+PROTO_BYTES_TYPES = (FieldDescriptorProtoType.TYPE_BYTES,)  # 12
 PROTO_MESSAGE_TYPES = (
-    FieldDescriptorProto.TYPE_MESSAGE,  # 11
-    FieldDescriptorProto.TYPE_ENUM,  # 14
+    FieldDescriptorProtoType.TYPE_MESSAGE,  # 11
+    FieldDescriptorProtoType.TYPE_ENUM,  # 14
 )
-PROTO_MAP_TYPES = (FieldDescriptorProto.TYPE_MESSAGE,)  # 11
+PROTO_MAP_TYPES = (FieldDescriptorProtoType.TYPE_MESSAGE,)  # 11
 PROTO_PACKED_TYPES = (
-    FieldDescriptorProto.TYPE_DOUBLE,  # 1
-    FieldDescriptorProto.TYPE_FLOAT,  # 2
-    FieldDescriptorProto.TYPE_INT64,  # 3
-    FieldDescriptorProto.TYPE_UINT64,  # 4
-    FieldDescriptorProto.TYPE_INT32,  # 5
-    FieldDescriptorProto.TYPE_FIXED64,  # 6
-    FieldDescriptorProto.TYPE_FIXED32,  # 7
-    FieldDescriptorProto.TYPE_BOOL,  # 8
-    FieldDescriptorProto.TYPE_UINT32,  # 13
-    FieldDescriptorProto.TYPE_SFIXED32,  # 15
-    FieldDescriptorProto.TYPE_SFIXED64,  # 16
-    FieldDescriptorProto.TYPE_SINT32,  # 17
-    FieldDescriptorProto.TYPE_SINT64,  # 18
+    FieldDescriptorProtoType.TYPE_DOUBLE,  # 1
+    FieldDescriptorProtoType.TYPE_FLOAT,  # 2
+    FieldDescriptorProtoType.TYPE_INT64,  # 3
+    FieldDescriptorProtoType.TYPE_UINT64,  # 4
+    FieldDescriptorProtoType.TYPE_INT32,  # 5
+    FieldDescriptorProtoType.TYPE_FIXED64,  # 6
+    FieldDescriptorProtoType.TYPE_FIXED32,  # 7
+    FieldDescriptorProtoType.TYPE_BOOL,  # 8
+    FieldDescriptorProtoType.TYPE_UINT32,  # 13
+    FieldDescriptorProtoType.TYPE_SFIXED32,  # 15
+    FieldDescriptorProtoType.TYPE_SFIXED64,  # 16
+    FieldDescriptorProtoType.TYPE_SINT32,  # 17
+    FieldDescriptorProtoType.TYPE_SINT64,  # 18
 )
 
 
+def monkey_patch_oneof_index():
+    """
+    The compiler message types are written for proto2, but we read them as proto3.
+    For this to work in the case of the oneof_index fields, which depend on being able
+    to tell whether they were set, we have to treat them as oneof fields. This method
+    monkey patches the generated classes after the fact to force this behaviour.
+    """
+    object.__setattr__(
+        FieldDescriptorProto.__dataclass_fields__["oneof_index"].metadata[
+            "betterproto"
+        ],
+        "group",
+        "oneof_index",
+    )
+    object.__setattr__(
+        Field.__dataclass_fields__["oneof_index"].metadata["betterproto"],
+        "group",
+        "oneof_index",
+    )
```

Comment:
Can you give me a rundown on why it needs to be done like this exactly? (I am trying to understand)

I looked up the help for this method

```
>>> help(object.__setattr__)
Help on wrapper_descriptor:

__setattr__(self, name, value, /)
    Implement setattr(self, name, value).
```

So my patchy current understanding is that the attribute `group` is being set to the string `"oneof_index"`

But I have no idea what is being passed to `self`

Reply:
Because it's being passed as the instance to the unbound method `__setattr__`

Reply:
@cetanu My attention was elsewhere so I only just noticed that you'd looked at this. thanks!

The [protoc messages](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/descriptor.proto#L62) are proto2 (and we only support proto3), so the [FieldDescriptorProto](https://github.com/danielgtaylor/python-betterproto/blob/master/src/betterproto/lib/google/protobuf/__init__.py#L468) and Field message types rely on proto2 `optional` semantics to know whether a these specific fields have been set.

Although proto3 doesn't support knowing whether a field has been set in this way, the binary for a proto2 optional field is compatible with a proto3 one_of with a single possible type!

Therefore we can replicate the intended proto2 semantics here with a betterproto message type if we force it to treat this field as if it were in a one_of group of its own, by setting the "group" value to something locally unique (like `"oneof_index"`).

This hack is nasty enough to call into question whether this whole endeavour of making betterproto use it's own message types internally is worthwhile given that it's aimed at proto3. However I think it's not that big a deal given that the blast radius is quite well contained, to the plugin process which I think is quite well exercised by the unit tests.

Reply:
The former is the part I didn't understand but now I do, the latter is interesting.
If we have some tests (doesn't necessarily have to be in this PR) that cover the proto2 and proto3 behavior that we want to support, then I see no issue with making this type of hack.

I'm okay to merge this